system/lib/pthread/emscripten_tls_init.c

```diff
@@ -21,15 +23,47 @@ extern void __wasm_init_tls(void *memory);
 
 extern int __dso_handle;
 
+// Create a thread-local wasm global which signal that the current thread is non
+// to use the primary/static TLS region.  Once this gets set it forces that all
+// future calls to emscripten_tls_init to dynamically allocate TLS.
+// This is needed because emscripten re-used module instances for new threads
+// which means stale values of __tls_base from a previous pthread need to be
+// ignored.
```

Comment:
Can we clear `__tls_base` when killing a thread instead? I guess not because it's only accessible via a builtin that only returns its value, right?

Reply:
I looked into that option.. but I feel like this one is more fool proof and cleaner (for now).

In the long run we are going to have to figure out how to free the TLS data of each side module (currently we leak that), so its possible we could combine the freeing with the clearing.  But for now I like this localized approach to the solving the problem which is fool proof and doesn't require a guaranteed clearing of the global.